reference/constraints/When.rst

```diff
@@ -0,0 +1,202 @@
+When
+===========
+
+This constraint allows you to apply constraints validation only if the provided expression returns true.
+See `Basic Usage`_ for an example.
+
+==========  ===================================================================
+Applies to  :ref:`class <validation-class-target>`
+            or :ref:`property/method <validation-property-target>`
+Options     - :ref:`expression <reference-constraint-expression-option>`
+            - :ref:`constraints <reference-constraint-constraints-option>`
+            - `groups`_
+            - `payload`_
+            - `values`_
+Class       :class:`Symfony\\Component\\Validator\\Constraints\\When`
+Validator   :class:`Symfony\\Component\\Validator\\Constraints\\WhenValidator`
+==========  ===================================================================
+
+Basic Usage
+-----------
+
+Imagine you have a class ``Discount`` with ``type`` and ``value``
+properties::
+
+    // src/Model/Discount.php
+    namespace App\Model;
+
+    use Symfony\Component\Validator\Constraints as Assert;
+
+    class Discount
+    {
+        private $type;
```

Comment:
you can leverage type hinting here and below (for better doc as sf6+ requires php8+)